### ISSUE TITLE
[mixed type] Supertype bug 

### ISSUE THREAD

**Issue by vieks** [facebook/flow#1644]
From the doc:

> The mixed type is a supertype of all types.

So:
```javascript
const foo: Object = {};
const bar: mixed = foo;
```
```No errors!```

OK, since mixed is the Supertype of 'Object'.

But:
```javascript
const foo: Array<Object> = [{}, {}];
const bar: Array<mixed> = foo;
```
```
144: const bar: Array<mixed> = foo;
                      ^^^^^ mixed. This type is incompatible with
143: const foo: Array<Object> = [{}, {}];
                      ^^^^^^ object type
```

WTF ??? Now mixed is not the supertype of Object ???

**Comment by samwgoldman** [comment]
This is expected behavior. Your assertion that `Array<B>` should be a subtype of `Array<A>` if `A` is a subtype of `B` isn't true. Phrased differently, your issue asserts that arrays are covariant. In fact, arrays are invariant.

Read-only arrays *would* be covariant, but since arrays are mutable in JS, they are invariant in Flow, which prevents unsound code like the following:

```js
class A {}
class B extends A {}
var bs: Array<B> = [];
var as: Array<A> = bs;
as.push(new A); // this would be bad!
```

**Comment by vieks** [comment]
OK thanks for the explanation, it make sense now.

But I never said this:

> Your assertion that Array<B> should be a subtype of Array<A> if A is a SUBtype of B isn't true.

Instead:
> Your assertion that Array<B> should be a subtype of Array<A> if A is a SUPERtype of B isn't true

**Comment by samwgoldman** [comment]
You're right—I misspoke. Thanks for the correction!

**Comment by vieks** [comment]
No prob.
You welcome :)

**Comment by vkurchatkin** [comment]
@samwgoldman thanks for the explanation! it seems to be the general rule for all the languages that only read-only types can be not invariant

**Comment by vieks** [comment]
Hum...

OK, so following your explanation:

```javascript
class Foo{}
class Bar extends Foo{}

const a: Array<Bar> = [];
const f: Array<Foo> = x;
```

```
164: const v: Array<Foo> = x;
                    ^^^ Foo. This type is incompatible with
163: const x: Array<Bar> = [];
                    ^^^ Bar
```
Alright ! Bar inherits from Foo, arrays are invariant, error is legitimate. 

But:

```javascript
const a: Array<String> = [];
const f: Array<Object> = x;
f.push({ a: 'dat fail :)' });// Holy hell !!!
```
```
No errors!
```

String inherits from Object but everything is OK ???
This is an inconsistent behavior with built-in Type ?

**Comment by samwgoldman** [comment]
Nice observation! The `Object` type is a supertype and subtype of all objects, similar to `any`. So `Array<Object>` is compatible with `Array<String>`, since `Object <: String` *and* `Object >: String`.

Aside, I wish we had named `Object` something like `AnyObject` instead, and made `Object` more specifically the type of `new Object` with methods like `toString` but nothing else.

**Comment by vieks** [comment]
Wow Indeed I could not even supposed ```Object``` acts as ```any``` (a kind of wildcard/bypass).

IMHO it's a bad behavior which mislead. I do not permit to emit a criticism you have made a great job with Flow, but please at least put the information in your website doc:

**Object**: The Object type is a supertype and subtype of all objects. That means Object is not the strict equivalent of the native/built-in Object type but is more similar to ```any```.
To sum up: ```any``` is the sub/supertype of all scalars and objects whilst ```Object``` is the sub/supertype of all Object only (except for the Array type). 

**Array**: IMPORTANT: Arrays are generics invariant.

Thanks for your support @samwgoldman.
